File: C++/Decorator/Decorator.cpp
#include "Decorator.h"
#include <iostream>

using namespace std;

Component::Component() {

}

Component::~Component() {

}

void Component::Operation() {

}

ConcreteComponent::ConcreteComponent() {

}

ConcreteComponent::~ConcreteComponent() {

}

void ConcreteComponent::Operation() { 
	std::cout << "ConcreteComponent operation..." << std::endl;
}

Decorator::Decorator(Component* com) { 
	this->_com = com;
}

Decorator::~Decorator() { 
	delete _com; 
}
void Decorator::Operation() {

}

ConcreteDecorator::ConcreteDecorator(Component*com) :Decorator(com) {

}

ConcreteDecorator::~ConcreteDecorator() {

}

void ConcreteDecorator::AddedBehavior() { 
	std::cout << "ConcreteDecorator::AddedBehacior...." << std::endl; 
} 

void ConcreteDecorator::Operation() {
	_com->Operation();
	this->AddedBehavior();
}

File: Java/Component/src/com/frank/Decorator.java
package com.frank;

/**
 * Created by frank on 2016/1/30.
 */
public abstract class Decorator  extends Component {
    protected Component component;

    public Decorator(Component component) {
        this.component = component;
    }

    public void operation() {
        // 在执行请求给component时执行其他命令
        System.out.println("Decorator operation");
        component.operation();
    }

}
